### PR TITLE
allow to specify absolute paths for reports. to support msbuild runner

### PR DESCRIPTION
**Pull request by jmecosta**
support absolute path of report paths

### CODE REVIEW

File: sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java [hunk 1]
```diff
@@ -187,6 +196,10 @@ protected String getStringProperty(String name, String def) {
         }
         
         for (String relPath : relPaths) {
+          CxxUtils.LOG.debug("Report '{}'", relPath);
+        }
+
```
Comment (guwirth) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: @jmecosta it's iterating twice over the paths now? Is there a reason for this?

Comment (jmecosta) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: will take this, left for debugging. forgot to remove

Comment (jmecosta) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: @guwirth would you be ok to test this?

**Comment by jmecosta** [comment]
comments done

**Comment by guwirth** [comment]
@jmecosta will try to test it this week

**Comment by guwirth** [comment]
Related to #425

**Comment by guwirth** [comment]
@jmecosta not sure if I understand the feature in the right way? Following it's not working;

**sonar-project.properties** at D:\Sonar\Test
```INI
sonar.cxx.other.reportPath=D:/xxx/reports/externalrules-*.xml
```
**log:**
```
23:31:54.840 INFO  - Base dir: D:\Sonar\Test
23:31:54.840 INFO  - Working dir: D:\Sonar\Test\.sonar
...
23:31:55.830 INFO  - Sensor CxxExternalRulesSensor
23:31:55.830 DEBUG - Using pattern 'D:\xxx\reports\externalrules-*.xml' to find reports
23:31:55.830 DEBUG - Search string transformed from 'D:\xxx\reports\externalrules-*.xml'
23:31:55.830 DEBUG - To ':\xxx\reports\externalrules-*.xml' to find reports, using base Path 'D:\Sonar\Test'
23:31:55.830 WARN  - Cannot find a report for 'sonar.cxx.other.reportPath=:\xxx\reports\externalrules-*.xml'
...
```

**Comment by jmecosta** [comment]
I will change it, don't think there any reasons not to use reports outside
project root. In this change you can only specify absolute paths that are
inside project root

**Comment by guwirth** [comment]
Still a problem:
```
09:54:01.296 INFO  - Sensor CxxExternalRulesSensor
09:54:01.296 DEBUG - Using pattern 'D:\xxx\reports\externalrules-*.xml' to find reports
09:54:01.296 DEBUG - Search string transformed from 'D:\xxx\reports\externalrules-*.xml'
09:54:01.296 DEBUG - To 'D:\xxx\reports\externalrules-*.xml' to find reports, using base Path 'D:\Sonar\Test'
09:54:01.296 WARN  - Cannot find a report for 'sonar.cxx.other.reportPath=D:\xxx\reports\externalrules-*.xml'
```

**Comment by guwirth** [comment]
@jmecosta Think would be better you write some unit tests. In my example above this should work:
```
 Base dir: D:\Sonar\Test
```
**inside root** absolute and relative inside root
```
sonar.cxx.other.reportPath=reports/externalrules-*.xml
sonar.cxx.other.reportPath=D:/Sonar/Test/reports/externalrules-*.xml
```
**outside root** absolute and relative outside root
```
sonar.cxx.other.reportPath=../../xxx/reports/externalrules-*.xml
sonar.cxx.other.reportPath=D:/xxx/reports/externalrules-*.xml
```
**outside root** on other drive
```
sonar.cxx.other.reportPath=C:/xxx/reports/externalrules-*.xml
```

File: sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java [hunk 2]
```diff
@@ -156,18 +156,32 @@ protected String getStringProperty(String name, String def) {
...
+          
+          CxxUtils.LOG.debug("To '{}' to find reports, using base Path '{}'", reportPath, baseNormalize);
+        }
+        
         DirectoryScanner scanner = new DirectoryScanner();
         String[] includes = new String[1];
         includes[0] = reportPath;
-        scanner.setIncludes(includes);
-        String baseDirPath = baseDirPath1;
         scanner.setBasedir(new File(baseDirPath));
```
Comment (guwirth) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: is old baseDirPath

File: sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java [hunk 3]
```diff
@@ -156,18 +156,32 @@ protected String getStringProperty(String name, String def) {
...
+        
+        if (singleFile.isAbsolute()) {    
+          String baseNormalize = FilenameUtils.normalize(baseDirPath);
+          CxxUtils.LOG.debug("Search string transformed from '{}'", reportPath);
+          reportPath = reportPath.replace(baseNormalize, "");
+          
+          if (reportPath.startsWith("/") || reportPath.startsWith("\\")) {
+            reportPath = reportPath.substring(1);
+          }
+
```
Comment (guwirth) [inline sonar-cxx-plugin/src/main/java/org/sonar/plugins/cxx/utils/CxxReportSensor.java]: baseDirPath should be updated here?

**Comment by jmecosta** [comment]
@guwirth i will write a integration test, i think its a better solution than unit test for this case. anyway this is still incomplete.  thanks for the initial  test, i will update and create the it and let you know its ready

**Comment by jmecosta** [comment]
@guwirth do you know why we have to basedir paths in the getReports function in CxxReportSensor? 

can get the use for 2 paths from the tests

**Comment by guwirth** [comment]
@jmecosta will test and if positive merge it this evening.

**Comment by guwirth** [comment]
@jmecosta I'm sorry but my example is still not working :-1: 
```
22:00:17.399 INFO  - Sensor CxxExternalRulesSensor
22:00:17.399 DEBUG - Using pattern 'D:\xxx\reports\externalrules-*.xml' to find reports
22:00:17.399 DEBUG - Unprocessed root directory 'D:\Sonar\Test'
22:00:17.399 DEBUG - Unprocessed report file 'D:\xxx\reports\externalrules-*.xml'
22:00:17.399 DEBUG - Processed root directory '\'
22:00:17.399 DEBUG - Processed report file 'D:\xxx\reports\externalrules-*.xml'
22:00:17.399 DEBUG - Unprocessed root directory 'D:\Sonar\Test'
22:00:17.399 DEBUG - Unprocessed report file 'D:\xxx\reports\externalrules-*.xml'
22:00:17.399 DEBUG - Processed root directory '\'
22:00:17.399 DEBUG - Processed report file 'D:\xxx\reports\externalrules-*.xml'
22:00:17.400 WARN  - Cannot find a report for 'sonar.cxx.other.reportPath=D:\xxx\reports\externalrules-*.xml'
22:00:17.400 INFO  - Sensor CxxExternalRulesSensor (done) | time=1ms
```

**Comment by jmecosta** [comment]
Ok suppose we need the it after all.

**Comment by jmecosta** [comment]
@guwirth actually it almost works. 

DEBUG: Processed report file 'e:\reports\reports-xunit\*.xml'
INFO: Cannot find a report for 'sonar.cxx.xunit.reportPath=e:\reports\reports-xunit\*.xml'
DEBUG: No reports found, nothing to process
INFO: Sensor CxxXunitSensor (done) | time=1ms
INFO: Sensor CxxCoverageSensor
DEBUG: Parsing coverage reports
INFO: Processing report 'e:\reports\bullseye-coverage-results\bullseyecoverage-result-0.xml'
INFO: Parsing 'Cobertura' format
INFO: Parsing 'Bullseye' format
INFO: Added report 'e:\reports\bullseye-coverage-results\bullseyecoverage-result-0.xml' (parsed by: BullseyeParser) to the coverage data

if path is real file then is able to read it. need to figure out how that directory scanner works with these cases.

**Comment by jmecosta** [comment]
@guwirth i am really surprised with the behaviour of this directory scanner... i wrote a test to check for existing paths... the log shows this:

18:58:10.808 [main] DEBUG CxxPlugin - Unprocessed root directory 'E:\src\data'
18:58:10.810 [main] DEBUG CxxPlugin - Unprocessed report file 'E:\reports\reports-rats\*.xml'
18:58:10.812 [main] DEBUG CxxPlugin - Processed root directory 'E:'
18:58:10.812 [main] DEBUG CxxPlugin - Processed report file 'reports\reports-rats\*.xml'

In test it asserts that
assertThat(relPaths.length).isEqualTo(1615);

When i run the analysis i get:

DEBUG: Using pattern 'E:\reports\reports-rats\*.xml' to find reports
DEBUG: Unprocessed root directory 'E:\src\data'
DEBUG: Unprocessed report file 'E:\reports\reports-rats\*.xml'
DEBUG: Processed root directory 'E:'
DEBUG: Processed report file 'reports\reports-rats\*.xml'
DEBUG: Number of Report Files Found '0'

Exactly the same output, one detects 1615 reports the other one 0.

What i am missing here? Can you cofirm this behaviour in your side? the test is this testWithRealDataInTest 

this solution should work, as per test. i am ensure why its not working

**Comment by jmecosta** [comment]
drop directory scanner and use 

http://wilddiary.com/list-files-matching-a-naming-pattern-java/

**Comment by guwirth** [comment]
@jmecosta did some tests removing reactor. Maybe you can take this into account => #695

**Comment by jmecosta** [comment]
are you suggesting to remove the reactor. for absolute paths the reactor does not apply, the base dir should be basepath or the padron is defined.

**Comment by jmecosta** [comment]
@guwirth i have something that works now for your case, hopefully it works also for other cases but the implementation is now totally different and tests need to be adjusted a lot.

also in my opinion we should isolate the tests for  path lookup, reports parsing and metrics saving (there should not be any unit tests really but integration tests), having unit tests that do all requires loads of maintenance if we need to refactor some parts as we are doing it now

wdyt?

**Comment by jmecosta** [comment]
the implementation is like this:

absolute Path: x:\abc\dfe\abc.xml
  . baseroot -> x:\abc\dfe\
  . search pattern -> abc.xml
  . recursive -> false

absolute Path: c:\abc\dfe\*.xml (or anytype of https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob)
  . baseroot -> x:\abc\dfe\
  . search pattern -> *.xml or anytype of glob
  . recursive -> false
 
relative Paths: abc.xml
  . baseroot -> Reactor base Path, or Module Base Path if not found in reactor
  . search pattern -> abc.xml
  . recursive -> false

relative Paths: *.xml (or anytype of https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob)
  . baseroot -> Reactor base Path, or Module Base Path if not found in reactor
  . search pattern -> *.xml (or anytype of https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob)
  . recursive -> false

relative Paths: **/abc.xml
  . baseroot -> Reactor base Path only because in fact all abc,xml in all folders will be found
  . search pattern -> abc.xml
  . recursive -> true

the code is much simpler now. also a notice the normalizepath function is really slow and we need to find some alternative also during this refactoring

**Comment by guwirth** [comment]
@jmecosta Travis build is working again. You can squash and rebase to get new results.

**Comment by jmecosta** [comment]
@guwirth its done, but this is not expected to pass for now. ive disable many unit tests. if we agree on a way to go then i will refactor the tests also

however i think you can test this to see if it supports your use case

**Comment by jmecosta** [comment]
@guwirth appveyour down?

**Comment by guwirth** [comment]
@jmecosta Travis & AppVeyor failing, problem in multi-module IT.

**Comment by SonarOpenCommunityAdmin** [comment]
appveyour should now pass. i need to check that it once it runs again.

**Comment by jmecosta** [comment]
@guwirth i added a new it test, its a complex multimodule project. it can be used to test the process of reports relative to module. its currently testing reports outside project folder, however you might need to create  a python function to change the contents of the reports if we are in windows or linux

**Comment by jmecosta** [comment]
@guwirth tests are now passing in windows, and its are also passing in all configurations. there is 1 test failing in unix only that i am ensure whats the issue. not sure why it fails. ive push a commit to improve the logging of the test. 

i supose review is ready, and please try also running with with your test data that was failing earlier

**Comment by jmecosta** [comment]
All green now

**Comment by guwirth** [comment]
@jmecosta it's working now :+1:  
@Bertk: do you also wanna test it?

* test with Windows 7 x64, SQ 5.1.2, this PR
* reports absolute outside of baseDir
* path in reports: absolute

```INI
sonar.cxx.other.reportPath=D:/xxx/reports/externalrules-*.xml
sonar.cxx.coverage.reportPath=D:/xxx/reports/coverage-*.xml
```

```XML
<?xml version="1.0"?>
<results>
<error file="D:\Sonar\Test\error_recovery\pi\Pi.cpp" line="11" ... />
<error file="d:\sonar\test\error_recovery\pi\pi.cpp" line="12" ... />
</results>
```

File: integration-tests/testdata/cpp-multimodule-project-2/A/2/sonar-project.properties [hunk 4]
```diff
@@ -0,0 +1,2 @@
+sonar.projectName=2
+sonar.sources=./
```
Comment (guwirth) [inline integration-tests/testdata/cpp-multimodule-project-2/A/2/sonar-project.properties]: Why is sources this ./ and not only .   ?

File: integration-tests/testdata/cpp-multimodule-project-2/B/sonar-project.properties [hunk 5]
```diff
@@ -0,0 +1,3 @@
+sonar.projectName=B
+sonar.sources=./
```
Comment (guwirth) [inline integration-tests/testdata/cpp-multimodule-project-2/B/sonar-project.properties]: Why is sources this ./ and not only .   ?

Comment (jmecosta) [inline integration-tests/testdata/cpp-multimodule-project-2/B/sonar-project.properties]: i think they are equivalent...

File: sonar-cxx-plugin/src/test/java/org/sonar/plugins/cxx/utils/CxxReportSensor_getReports_Test.java [hunk 6]
```diff
@@ -75,12 +75,9 @@ public void getReports_patternMatching() throws java.io.IOException, java.lang.I
 
     examples.add(new String[] { "A?.ext",       "AA.ext,AB.ext",   "B.ext" });            // containing question mark
     examples.add(new String[] { "A*.ext",       "A.ext,AAA.ext",   "B.ext" });            // containing question mark
-    examples.add(new String[] { "**/A.ext",     "A.ext,dir/A.ext", "B.ext" });            // containing question mark
```
Comment (jmecosta) [inline sonar-cxx-plugin/src/test/java/org/sonar/plugins/cxx/utils/CxxReportSensor_getReports_Test.java]: @guwirth this case is failing on linux, windows it passes just fine. however this is a special case for recursive search wich should be handled. when i have access to a linux box i will double check why it fails.

**Comment by guwirth** [comment]
@jmecosta could you squash and rebase that please that we can merge it.
@bertk any comments from your side?

**Comment by jmecosta** [comment]
done

**Comment by Bertk** [comment]
@jmecosta I started an analysis with the PR this morning and results are OK. By the way I also used PR #695 :wink:

**Comment by guwirth** [comment]
@Bertk what is your main use case:
* path to reports in config file relative or absolute?
* reports inside or outside baseDir?
* file items in report files relative or absolute?

**Comment by Bertk** [comment]
@guwirth  
* I use relative and absolute paths for the reports
* additional I use multiple test reports for one component ( comma separate) . The ant pattern is useless because the files located in different folders and I use absolute paths for the test results
* the reports for pc-lint, cppcheck, Vera++, RATS are located in basedir. I use post-build events to create this reports within CI builds
* I use a mixture of relative or absolute paths for the files within the reports (vc compiler and pc-lint with absolute paths, Vera++ and CppCheck or RATS with relative paths)

**Pull request merged**